ls1intum/Artemis
# `Development`: Modernize appearance of quiz statistics

### Checklist
#### General


#### Client

### Motivation and Context


This PR is changing the underlying framework for the charts located in quiz statistics from ng2-charts/chart.js to ngx-charts.

### Description

The html templates as well as the underlying logic in the components are refactored in order to fit the requests of ngx-charts.
### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 ( or more) Students
- 1 Quiz containing at least one question of each possible type (MC, DnD, SA)

**Important**: I recommend testing this PR the following way: 
1. Do the steps below one time **before** deploying this PR. 
2. Inspect the old implementation. Maybe take screenshots in order to make the comparison a bit easier.
3. Deploy the PR and re-inspect the charts with the new implementation. 

Alternatively, you can inspect [these statistics](https://artemistest.ase.in.tum.de/course-management/481/quiz-exercises/13559/quiz-point-statistic) on TS1, e.g. with test user 2.

Here are the "original" steps for testing:

1. Log in to Artemis
2. Navigate to Course Administration
3. Set the Quiz visible and start it
4. Participate with the students and submit different answers for each exercise
5. Wait until the working time is finished
6. Switch to the instructor again
7. Navigate to Course Administration
8. In the exercises overview, in the row of the quiz, click on the "Statistics" button
9. Click through the different charts and make sure the displayed information did not change. Also check the behavior when several buttons (e.g. "Show sample solution" or "Show unrated results") are clicked.

Two last comments: 
- I know, the data labels are angled. I know, some of you do not like this. The reason why I decided to leave them angled is the appearance on small screens. Without that angle, the data labels overlap sooner or later. I think it is a more robust solution this way. But feel free to share your opinion on that
- You might notice that the appearance of these bar charts differ a bit from other implementations with `ngx-charts`. The edges of the bars are not rounded here. This is due to an error output I was facing while testing this implementation locally (to be more precise, this error only occurs when the "show (un-)rated results" button is triggered). `d3` seems to have issues when it comes to changing bars representing a non-zero value to zero. It is a bit weird because this error occurred for all charts except the point distribution chart (where the value 0 occurs as well). I did some research and found [this issue thread](https://github.com/swimlane/ngx-charts/issues/498). There, disabling the edge rounding was proposed as workaround, which worked for me. 

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage






### Screenshots


Old design of quiz point distribution:
![image](https://user-images.githubusercontent.com/80622272/145713292-26e07522-c1e6-4517-9880-b65641612af2.png)

**New** design of quiz point distribution:
![image](https://user-images.githubusercontent.com/80622272/145712970-a7392d01-4840-48b8-bac7-80cca434d564.png)


Old design of Quiz Statistics:
![image](https://user-images.githubusercontent.com/80622272/145712002-4e7807ac-c5e0-4e4a-b7ec-61280e712cdf.png)

**New** design of Quiz Statistics:
![image](https://user-images.githubusercontent.com/80622272/145712436-478428c4-7abb-42af-ad75-b9f1994e289a.png)


Old design of MC question statistics:
![image](https://user-images.githubusercontent.com/80622272/145712180-4b58a3f7-e5e6-4c6c-a2c2-5ea431011dd0.png)

**New** design of MC question statisitcs:
![image](https://user-images.githubusercontent.com/80622272/145712461-3b733701-762d-4fb7-a84f-fb64e444b439.png)


Old design of DnD question statistics:
![image](https://user-images.githubusercontent.com/80622272/145712191-b0ffa17b-b01d-48da-8377-08133b5a9c3d.png)

**New** design of DnD question statisitcs:
![image](https://user-images.githubusercontent.com/80622272/145712487-32204bc8-f48d-42bc-bec1-912cccee2902.png)


Old design of SA question statistics:
![image](https://user-images.githubusercontent.com/80622272/145712213-6738534c-bced-46b1-b223-ba332eec3ccb.png)

**New** design of SA question statisitcs:
![image](https://user-images.githubusercontent.com/80622272/145712517-1eaae986-11c7-4fa0-8506-0f93878d949c.png)
